rnd/autogpt_server/autogpt_server/executor/manager.py

```diff
@@ -350,76 +354,111 @@ def on_node_execution(cls, q: ExecutionQueue[NodeExecution], data: NodeExecution
     @classmethod
     def on_graph_executor_start(cls):
         cls.pool_size = Config().num_node_workers
-        cls.executor = ProcessPoolExecutor(
-            max_workers=cls.pool_size,
+        cls._init_node_executor_pool()
+        logger.warning(f"Graph executor started with max-{cls.pool_size} node workers.")
+
+    @classmethod
+    def _init_node_executor_pool(cls):
+        cls.executor = Pool(
+            processes=cls.pool_size,
             initializer=cls.on_node_executor_start,
         )
-        logger.warning(f"Graph executor started with max-{cls.pool_size} node workers.")
 
     @classmethod
-    def on_graph_execution(cls, graph_data: GraphExecution):
+    def on_graph_execution(cls, graph_data: GraphExecution, cancel: threading.Event):
         prefix = get_log_prefix(graph_data.graph_exec_id, "*")
         logger.warning(f"{prefix} Start graph execution")
 
+        finished = False
+
+        def cancel_handler():
+            while not cancel.is_set():
+                cancel.wait(1)
+            if finished:
+                return
+            cls.executor.terminate()
+            logger.info(
+                f"{prefix} Terminated graph execution {graph_data.graph_exec_id}"
+            )
+            cls._init_node_executor_pool()
+
+        cancel_thread = threading.Thread(target=cancel_handler)
```

Comment:
Can we reuse the thread on avoiding creating thread overhead on each execution ?

Reply:
I don't think so. Afaics we need to re-create and pass a new cancel `Event` for every execution, and `cancel_handler()` (partially) relies on the continuity of the `Event` object.

Reply:
Hmm this is definitely a very expensive process as opposed on using a threadpool to re-use it. We might need to revisit this as part of the performance improvement.

Reply:
I'm not sure how you envision this working with a threadpool. Since you seem to have a better idea, could you create a ticket?